### ISSUE TITLE
[Bug] [Tune] [1.8.0] [1.7.0] Race condition in cluster checkpoint syncing

### ISSUE THREAD

**Issue by dmkulazhenko** [ray-project/ray#20183]
## Abstract
A lot of checkpoints are not syncing from workers to head (driver) during tune job.

## Summary
Imagine that we have common `tune.Experiment` running via `tune.run`. All this executes in AWS using Ray Cluster Launcher.

#### Steps of execution
1. Cluster setups using `ray up`;
2. Payload submits using `ray submit`;
3. Right after end of execution `ray submit` -> `ray down` triggers.

#### Expected
1. Head node distributes tuner jobs over worker nodes (we doesn't compute anything on head: `{"cpu": 0, "gpu": 0}`;
2. Worker nodes complete jobs and save info about them (results, params, checkpoints) to worker disk (lets name this directory with info - `~/experiments`);
3. `NodeSyncer` syncs (using `rsync`) `~/experiments` to head node;
4. `CloudSyncer` syncs (using aws tools) `~/experiments` to S3 bucket. 

#### Actual
`NodeSyncer` syncs (using `rsync`) `~/experiments` to head node, but `checkpoint_xxxx` remains on worker node.

Example of head node `~/experiments/<exp_name>/Tuner_xxxx`:
```bash
[ray-master@ip-10-111-2-206 Tuner_d409ca08]$ ls -la
total 64
drwxrwxr-x   2 ray-master ray-master  4096 Nov  8 13:43 .
drwxrwxr-x 202 ray-master ray-master 12288 Nov  8 13:44 ..
-rw-rw-r--   1 ray-master ray-master  1550 Nov  8 13:44 events.out.tfevents.1636378997.ip-10-111-2-206.eu-west-1.compute.internal
-rw-------   1 ray-master ray-master   120 Nov  8 13:44 log_sync_outjo8h6p6f.log
-rw-rw-r--   1 ray-master ray-master  4742 Nov  8 13:43 params.json
-rw-rw-r--   1 ray-master ray-master 19170 Nov  8 13:43 params.pkl
-rw-rw-r--   1 ray-master ray-master   477 Nov  8 13:44 progress.csv
-rw-rw-r--   1 ray-master ray-master  5280 Nov  8 13:44 result.json
```

Example of worker node `~/experiments/<exp_name>/Tuner_xxxx`:
```bash
[ray-master@ip-10-111-1-139 Tuner_d409ca08]$ ls -la
total 12
drwxrwxr-x  3 ray-master ray-master 4096 Nov  8 13:44 .
drwxrwxr-x 26 ray-master ray-master 4096 Nov  8 13:43 ..
drwxrwxr-x  2 ray-master ray-master 4096 Nov  8 13:44 checkpoint_000001
```

#### `SyncConfig`
* `upload_dir` - never changed, just a S3 URI;
* `sync_to_cloud` - default (never passed);
* `sync_to_driver` - True / unset;
* `sync_on_checkpoint` - True / False;
* `node_sync_period` - 1 / 5 / 60 / 300;
* `cloud_sync_period` - 5 / 60 / 180 / 300;

All changes with sync config doesn't noticeably influence on described problem;

#### Service degradation
Seems to be we don't have such problems with `ray==1.5.2` and they come only with bumping to `ray==1.7.0`. 
I'm not really sure with that statement, because this bug is floating as we will see bellow.

#### Update to `ray==1.8.0`
Same as `ray==1.7.0`.

#### Cluster size / tuner jobs 
With some huge tuner jobs and low-size cluster (when whole experiment takes more than 2h) - syncing works partially OK.
And here things become more interesting :smiley: 

##### Run 1
* Cluster
  * Head: 1 x m5.large
  * Worker: 12 x c5n.4xlarge (192 cores - 192 workers in parallel)
* Job
  * 2000 samples (~11 samples per one core)
  * 1 sample ~ 5m (overall ~1h)

##### Run 2
* Cluster
  * Head: 1 x m5.large
  * Worker: 11 x c5n.4xlarge (176 cores - 176 workers in parallel)
* Job
  * 160 samples (<1 samples per one core)
  * 1 sample ~40s (overall ~1m)

##### Run 3
* Cluster
  * Head: 1 x m5.large
  * Worker: 10 x c5n.4xlarge (160 cores - 160 workers in parallel)
* Job
  * 200 samples (1-2 samples per one core)
  * 1 sample ~40s (overall ~2m)

##### Results
* Run 1: >80% checkpoints in S3;
* Run 2: no checkpoints in S3;
* Run 3: 2 checkpoints in S3.

##### Interesting artifact of run 3
Lets try to find any checkpoints on head node on run 3:
```bash
[ray-master@ip-10-111-2-206 kulazhenko_checkpoint_10]$ ls -la ./Tuner*/checkpoint*/
./Tuner_9fdcd842/checkpoint_000001/:
total 6980
drwxrwxr-x 2 ray-master ray-master    4096 Nov  8 13:42 .
drwxrwxr-x 3 ray-master ray-master    4096 Nov  8 13:42 ..
-rw-rw-r-- 1 ray-master ray-master 7137030 Nov  8 13:42 checkpoint.p
-rw-rw-r-- 1 ray-master ray-master       0 Nov  8 13:42 .is_checkpoint

./Tuner_ac1bc85c/checkpoint_000001/:
total 9144
drwxrwxr-x 2 ray-master ray-master    4096 Nov  8 13:42 .
drwxrwxr-x 3 ray-master ray-master    4096 Nov  8 13:42 ..
-rw-rw-r-- 1 ray-master ray-master 9351702 Nov  8 13:42 checkpoint.p
-rw-rw-r-- 1 ray-master ray-master       0 Nov  8 13:42 .is_checkpoint
[ray-master@ip-10-111-2-206 kulazhenko_checkpoint_10]$ ls -la ./Tuner_9fdcd842/
total 68
drwxrwxr-x   3 ray-master ray-master  4096 Nov  8 13:42 .
drwxrwxr-x 202 ray-master ray-master 12288 Nov  8 13:44 ..
drwxrwxr-x   2 ray-master ray-master  4096 Nov  8 13:42 checkpoint_000001
-rw-rw-r--   1 ray-master ray-master  1550 Nov  8 13:42 events.out.tfevents.1636378910.ip-10-111-2-206.eu-west-1.compute.internal
-rw-------   1 ray-master ray-master   224 Nov  8 13:42 log_sync_outqmdpfarr.log
-rw-rw-r--   1 ray-master ray-master  4743 Nov  8 13:41 params.json
-rw-rw-r--   1 ray-master ray-master 18927 Nov  8 13:41 params.pkl
-rw-rw-r--   1 ray-master ray-master   493 Nov  8 13:42 progress.csv
```

#### Exclude
I saw, that from 1.5.2 to 1.7.0 in sync-logic added `exclude` KW, but as far as sometimes checkpoints go to right place -- problem not in `exclude`.

#### Race condition
Seems to be:
* `rsync` executes before checkpoint even exists in worker folder;

or

* Synchronization isn't fully completed and killed by ray, because tune jobs completed. Also, keep in mind, that I kill cluster right after `ray submit` completes.

#### Cluster config
```yaml
auth:
  ssh_private_key: /home/ray/ssh-private-key.pem
  ssh_user: ray-master
available_node_types:
  test_head_3bcd38b8-6169-403f-91cb-12daf1573ba1:
    max_workers: 0
    min_workers: 0
    node_config:
      BlockDeviceMappings:
      - DeviceName: /dev/sda1
        Ebs:
          VolumeSize: 100
      IamInstanceProfile:
        Name: RAT-cluster
      ImageId: ami-013ddaa1a10a512f4
      InstanceType: m5.large
      KeyName: dkulazhenko-calib-test
      SecurityGroupIds:
      - sg-071cb5f0abb5d8c39
      SubnetId: subnet-0cbd17fdeaff1cfa8
    resources:
      CPU: 0
  test_worker:
    max_workers: 10
    min_workers: 10
    node_config:
      BlockDeviceMappings:
      - DeviceName: /dev/sda1
        Ebs:
          VolumeSize: 50
      IamInstanceProfile:
        Name: RAT-cluster
      ImageId: ami-013ddaa1a10a512f4
      InstanceType: c5n.4xlarge
      KeyName: dkulazhenko-calib-test
      SecurityGroupIds:
      - sg-071cb5f0abb5d8c39
      SubnetId: subnet-0cbd17fdeaff1cfa8
    resources: {}
cluster_name: rat_test_3bcd38b8-6169-403f-91cb-12daf1573ba1
file_mounts:
  /home/ray-master/git_ssm_key.sh: /home/ray/git_ssm_key.sh
head_node_type: test_head_3bcd38b8-6169-403f-91cb-12daf1573ba1
idle_timeout_minutes: 1
initialization_commands:
- if [ $(sudo file -s /dev/nvme1n1 | cut -d ':' -f 2 | tr -d ' ') = 'data' ]; then
  sudo sudo mkfs -t ext4 /dev/nvme1n1; else echo 'ext4 partition already exists on
  /dev/nvme1n1'; fi
- mount | awk '{if ($3 == "/home/ray-master/pipeopt_experiments") { exit 0}} ENDFILE{exit
  1}' || mkdir /home/ray-master/pipeopt_experiments && sudo mount /dev/nvme1n1 /home/ray-master/pipeopt_experiments
  && sudo chown -R ray-master:ray-master /home/ray-master/pipeopt_experiments && sudo
  chmod -R 775 /home/ray-master/pipeopt_experiments
max_workers: 10
provider:
  availability_zone: eu-west-1a
  cache_stopped_nodes: false
  region: eu-west-1
  type: aws
setup_commands:
- test -e /home/ray-master/analytical-pipelines || /home/ray-master/git_ssm_key.sh
  test_param clone --depth=1 -b v1.1.0 git@ssh.dev.azure.com:v3/skfdc/REP-SW/analytical-pipelines
- test -e /home/ray-master/pyimdb || /home/ray-master/git_ssm_key.sh git-ssh-key clone
  git@ssh.dev.azure.com:v3/skfdc/REP-SW/pyimdb
- pip install /home/ray-master/analytical-pipelines[all]
- pip install /home/ray-master/pyimdb
- pip install mlflow==1.20.1
upscaling_speed: 100
```

#### Payload submitted via `ray submit`
There is a lot of useless stuff for you, but I'm trying to give you a whole picture :smiley: 
`Driver`, `SKFExperimentAPI` and so on - our internal objects that transform our pipelines to `tune.Experiment` and execute them via `tune.run`.
```python
import argparse
import json
import time
import traceback

import boto3
from enum import Enum
from mlflow.entities import Experiment, RunStatus
from mlflow.tracking import MlflowClient
from pipeline.api.skf_experiment import SKFExperimentAPI
from pipeline.pipeopt import Driver
from pipeline.pipeopt.callbacks import PipeoptMLflowLoggerCallback
from pipeline.utils.aws import get_param
from ray.tune import SyncConfig

# I deleted functions related to statuses updates because they are 100% useless.

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("config_path")
    args = parser.parse_args()
    config_path = args.config_path

    buckets = json.loads(get_param("auto-ml-experiment-buckets"))
    config_bucket, artifacts_bucket = (
        buckets["config_bucket"],
        buckets["artifacts_bucket"],
    )

    s3_resource = boto3.resource("s3")
    config = json.loads(
        s3_resource.Object(config_bucket, config_path)
        .get()["Body"]
        .read()
        .decode("utf-8")
    )
    mlflow_tracking_uri = json.loads(get_param("mlflow_configs"))[
        "tracking_uri"
    ]
    mlflow_client: MlflowClient = MlflowClient(tracking_uri=mlflow_tracking_uri)
    mlflow_experiment = get_experiment(mlflow_client, config.get("name"))
    mlflow_experiment_id = mlflow_experiment.experiment_id
    # noinspection PyBroadException
    try:
        experiment = SKFExperimentAPI(**config).instantiate()
        experiment.callbacks = [
            PipeoptMLflowLoggerCallback(
                experiment_name=config.get("name"),
                tracking_uri=mlflow_tracking_uri,
            )
        ]
        experiment.sync_config = SyncConfig(
            upload_dir=f"s3://<bucket_name>/experiments/{config.get('name')}/trials",
            sync_to_driver=True,
            node_sync_period=5,
            cloud_sync_period=30
        )
    except Exception:
        traceback.print_exc()
        set_experiment_status(mlflow_client, mlflow_experiment_id, ExperimentStatus.FAILED)
        return 1

    driver = Driver(ray_init_params={"address": "auto"})
    set_experiment_status(mlflow_client, mlflow_experiment_id, ExperimentStatus.RUNNING)
    # noinspection PyBroadException
    try:
        driver.execute(experiment)
        set_experiment_status(mlflow_client, mlflow_experiment_id, ExperimentStatus.COMPLETED)
    except Exception:
        traceback.print_exc()
        fail_unfinished_runs(mlflow_client, mlflow_experiment_id)
        set_experiment_status(mlflow_client, mlflow_experiment_id, ExperimentStatus.FAILED)
        return 1


if __name__ == "__main__":
    main()
``` 

#### Cluster managing
We upping / downing cluster using our internal wrapper of ray CLI and it looks smth like this.
```python
import json
import os
from copy import deepcopy
from functools import lru_cache
from typing import Dict, Optional
from uuid import uuid4

import boto3
from ray_auto_tune.cluster import ClusterMaster
from ray_auto_tune.config.cluster import ClusterConfig
from ray_auto_tune.config.node import NodeConfig
from ray_auto_tune.utils import get_logger
from urllib.request import pathname2url

# deleted a lot of useless stuff

def main():
    bucket_name = os.environ.get("BUCKET_NAME")
    config = get_cluster_config(bucket_name)
    save_config_to_s3(config, bucket_name)

    cluster = ClusterMaster(config, execution_id=os.environ.get("CALIBRATION_EXPERIMENT_NAME"))
    logger = get_logger(__name__)

    logger.debug("Started cluster up.")
    _, error, return_code = cluster.up(output_to_std=False)
    if return_code != 0:
        logger.debug(
            f"Cluster up failed with return code {return_code}! Error:\n {error}"
        )
        return
    logger.debug("Finished cluster up. Started submitting an experiment.")

    config_path = os.environ.get("CALIBRATION_CONFIG_PATH")
    logger.debug(f"Calibration config path: {config_path}. Bucket name: {bucket_name}")

    try:
        # we need pathname2url in case of spaces in experiment name
        return_code = cluster.submit("experiment.py", pathname2url(config_path), bucket_name)
        if return_code != 0:
            logger.debug(
                f"Experiment failed with an error and return code {return_code}!"
            )
    except Exception as e:
        logger.debug(e)

    cluster.down()
```

#### Traces
```
2021-11-08T13:03:53.308Z,,ray-auto-tune,"2021-11-08 13:00:37,179	ERROR syncer.py:192 -- Sync execution failed."
2021-11-08T13:03:53.383Z,,ray-auto-tune,Traceback (most recent call last):
2021-11-08T13:03:53.508Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 188, in sync_down"
2021-11-08T13:03:53.585Z,,ray-auto-tune,"result = self.sync_client.sync_down(self._remote_path,"
2021-11-08T13:03:53.665Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 214, in sync_down"
2021-11-08T13:03:53.740Z,,ray-auto-tune,"return self._execute(self.sync_down_template, source, target)"
2021-11-08T13:03:53.810Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 273, in _execute"
2021-11-08T13:03:53.913Z,,ray-auto-tune,stdout=self._get_logfile())
2021-11-08T13:03:53.984Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 203, in _get_logfile"
2021-11-08T13:03:54.033Z,,ray-auto-tune,raise RuntimeError(
2021-11-08T13:03:54.114Z,,ray-auto-tune,RuntimeError: [internalerror] The client has been closed. Please report this stacktrace + your cluster configuration on Github!
2021-11-08T13:03:54.192Z,,ray-auto-tune,"2021-11-08 13:00:37,181	ERROR syncer.py:415 -- Trial Tuner_c9c38774: Checkpoint sync skipped. This should not happen."
2021-11-08T13:03:54.259Z,,ray-auto-tune,"2021-11-08 13:00:37,182	ERROR trial_runner.py:1013 -- Trial Tuner_c9c38774: Error handling checkpoint /home/ray-master/pipeopt_experiments/kulazhenko_checkpoint_9/Tuner_c9c38774/checkpoint_000001/"
2021-11-08T13:03:54.322Z,,ray-auto-tune,Traceback (most recent call last):
2021-11-08T13:03:54.370Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/trial_runner.py"", line 1005, in _process_trial_save"
2021-11-08T13:03:54.438Z,,ray-auto-tune,self._callbacks.on_checkpoint(
2021-11-08T13:03:54.509Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/callback.py"", line 259, in on_checkpoint"
2021-11-08T13:03:54.539Z,,ray-auto-tune,callback.on_checkpoint(**info)
2021-11-08T13:03:54.615Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 457, in on_checkpoint"
2021-11-08T13:03:54.676Z,,ray-auto-tune,"self._sync_trial_checkpoint(trial, checkpoint)"
2021-11-08T13:03:54.706Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 430, in _sync_trial_checkpoint"
2021-11-08T13:03:54.792Z,,ray-auto-tune,"raise TuneError(""Trial {}: Checkpoint path {} not """
2021-11-08T13:03:54.867Z,,ray-auto-tune,ray.tune.error.TuneError: Trial Tuner_c9c38774: Checkpoint path /home/ray-master/pipeopt_experiments/kulazhenko_checkpoint_9/Tuner_c9c38774/checkpoint_000001/ not found after successful sync down.
```

```
2021-11-08T13:03:59.363Z,,ray-auto-tune,"2021-11-08 13:00:39,605	ERROR syncer.py:192 -- Sync execution failed."
2021-11-08T13:03:59.420Z,,ray-auto-tune,Traceback (most recent call last):
2021-11-08T13:03:59.470Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 188, in sync_down"
2021-11-08T13:03:59.541Z,,ray-auto-tune,"result = self.sync_client.sync_down(self._remote_path,"
2021-11-08T13:03:59.619Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 214, in sync_down"
2021-11-08T13:03:59.695Z,,ray-auto-tune,"return self._execute(self.sync_down_template, source, target)"
2021-11-08T13:03:59.758Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 273, in _execute"
2021-11-08T13:03:59.824Z,,ray-auto-tune,stdout=self._get_logfile())
2021-11-08T13:03:59.905Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py"", line 203, in _get_logfile"
2021-11-08T13:03:59.971Z,,ray-auto-tune,raise RuntimeError(
2021-11-08T13:04:00.030Z,,ray-auto-tune,RuntimeError: [internalerror] The client has been closed. Please report this stacktrace + your cluster configuration on Github!
2021-11-08T13:04:00.105Z,,ray-auto-tune,"2021-11-08 13:00:39,605	ERROR syncer.py:415 -- Trial Tuner_c869cbd6: Checkpoint sync skipped. This should not happen."
2021-11-08T13:04:00.183Z,,ray-auto-tune,"2021-11-08 13:00:39,606	ERROR trial_runner.py:1013 -- Trial Tuner_c869cbd6: Error handling checkpoint /home/ray-master/pipeopt_experiments/kulazhenko_checkpoint_9/Tuner_c869cbd6/checkpoint_000001/"
2021-11-08T13:04:00.265Z,,ray-auto-tune,Traceback (most recent call last):
2021-11-08T13:04:00.337Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/trial_runner.py"", line 1005, in _process_trial_save"
2021-11-08T13:04:00.411Z,,ray-auto-tune,self._callbacks.on_checkpoint(
2021-11-08T13:04:00.530Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/callback.py"", line 259, in on_checkpoint"
2021-11-08T13:04:00.625Z,,ray-auto-tune,callback.on_checkpoint(**info)
2021-11-08T13:04:00.682Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 457, in on_checkpoint"
2021-11-08T13:04:00.756Z,,ray-auto-tune,"self._sync_trial_checkpoint(trial, checkpoint)"
2021-11-08T13:04:00.828Z,,ray-auto-tune,"File ""/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py"", line 430, in _sync_trial_checkpoint"
2021-11-08T13:04:00.894Z,,ray-auto-tune,"raise TuneError(""Trial {}: Checkpoint path {} not """
2021-11-08T13:04:00.963Z,,ray-auto-tune,ray.tune.error.TuneError: Trial Tuner_c869cbd6: Checkpoint path /home/ray-master/pipeopt_experiments/kulazhenko_checkpoint_9/Tuner_c869cbd6/checkpoint_000001/ not found after successful sync down.
```

This one is pretty rare, I found it accidentally :smiley: 
```
2021-11-08 00:19:50,649	ERROR syncer.py:210 -- Sync execution failed.
Traceback (most recent call last):
  File "/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/syncer.py", line 206, in sync_down
    result = self.sync_client.sync_down(
  File "/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py", line 264, in sync_down
    return self._execute(self.sync_down_template, source, target, exclude)
  File "/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py", line 347, in _execute
    stdout=self._get_logfile())
  File "/home/ray-master/.pyenv/versions/3.8.11/lib/python3.8/site-packages/ray/tune/sync_client.py", line 253, in _get_logfile
    raise RuntimeError(
RuntimeError: [internalerror] The client has been closed. Please report this stacktrace + your cluster configuration on Github!
2021-11-08 00:19:50,649	ERROR syncer.py:440 -- Trial Tuner_74cbcc46: Checkpoint sync skipped. This should not happen.
2021-11-08 00:19:50,649	WARNING util.py:165 -- The `process_trial_save` operation took 0.647 s, which may be a performance bottleneck.
Trial Tuner_737c450a reported f1=0.69 with parameters={'BeforeDowntimeFullPeriodLabeler1__pred_interval': 127, 'RandomForestRFE1__n_features_to_select': 143, 'RandomForestClassifier1__n_estimators': 137, 'RandomForestClassifier1__min_samples_split': 187, 'HalfGaussRVFilter1__kernel_size': 139, 'ThresholdOptimizer1__compromise': 0.017920894478876504}. This trial completed.
```

And I have a full log of calibration similar to run 3: [extract-2021-11-09_14-11-86.csv](https://github.com/ray-project/ray/files/7505051/extract-2021-11-09_14-11-86.csv)

## Similar issues
I found #13256. And seems to be it's pretty similar, but last time @krfricke cannot reproduce it... And, probably, I know why - his run was similar to my run 1.

------

Thank you in advance, if there will be something unclear or you will need more information - feel free to ask.

**Comment by krfricke** [comment]
Thanks for raising this - we are in the process of refactoring cloud checkpointing and simplifying configuration (e.g. https://github.com/ray-project/ray/pull/20155).

For your use case, could you try using an S3 bucket for cloud checkpointing? This is the recommended solution for distributed training on the cloud (e.g. AWS) anyways.

**Comment by dmkulazhenko** [comment]
@krfricke, sorry, I don't get it :crying_cat_face: 

I meant, that sync \<workers\> -- \<head\> doesn't work in right way... But with sync \<head\> -- \<S3\> all is OK.

Im already using S3 bucket as final place for checkpoint -- you can check my `upload_dir`, which is S3 URI.
So, to "hack" my problem - I just upload checkpoints to S3 directly from `tune.Trainable.save_checkpoint`.

**Comment by krfricke** [comment]
Ah right, sorry. I read through the issue once more and I think I understand your problem now.

Let me recap just to be sure:

- You have trials running on worker nodes
- The checkpoints should be synced back to the head node (via a NodeSyncer)
- From there, they should be synced to S3 (via the ClusterSyncer)

To do this in the latest release you have to use `tune.durable(trainable)` to wrap around your trainable. This will enable automatic trial checkpointing to the cloud.
Please note that this will disable checkpoint syncing to the driver - checkpoints will be directly synced from the remote trials to the s3 bucket. It seems like this is the solution you implemented yourself in your trainable.

We just got rid of this extra step in this PR: https://github.com/ray-project/ray/pull/19880 which will be included in the next release (1.9.0). We are also very close to landing https://github.com/ray-project/ray/pull/20155 which will refactor and simplify the configuration interface.

Let me know if this solves your problem!
